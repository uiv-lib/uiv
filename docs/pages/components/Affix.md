# Affix

> The affix component toggles `position: fixed` on and off, emulating the effect found with `position: sticky`.

The navigation **on the right** is a live demo of the affix component (hidden on mobile devices).

**Note**:

* You must provide CSS for the positioning and width of your affixed content.
* Do not use affix on an element contained in a relatively positioned element, such as a pulled or pushed column.

<ins class="adsbygoogle"
     style="display:block; text-align:center;"
     data-ad-layout="in-article"
     data-ad-format="fluid"
     data-ad-client="ca-pub-4714899946256166"
     data-ad-slot="4603582855"></ins>

## Example

```html
<affix :offset="50">
  <ul>
    <li>...</li>
    <li>...</li>
    <li>...</li>
  </ul>
</affix>
```

# API Reference

## [Affix](https://github.com/wxsms/uiv/blob/master/src/components/affix/Affix.vue)

### Props

Name             | Type       | Default      | Required | Description
---------------- | ---------- | ------------ | -------- | -----------------------
`offset`         | Number     | 0            |          | Pixels to offset from screen when calculating position of scroll.

### Slots

Name      | Description
--------- | -----------------------
`default` | The affix body.

### Events

Name      | Description
--------- | -----------------------
`affix`   | This event fires immediately before the element has been affixed.
`affixed` | This event is fired after the element has been affixed.
