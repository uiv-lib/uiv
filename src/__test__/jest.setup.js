import { enableAutoDestroy } from '@vue/test-utils'

enableAutoDestroy(afterEach)

window.scrollTo = (x, y) => {
  document.documentElement.scrollTop = y
  document.documentElement.scrollLeft = x
  window.pageYOffset = y
  window.pageXOffset = x
  window.dispatchEvent(new Event('resize'))
}
