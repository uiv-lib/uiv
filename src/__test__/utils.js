import Vue from 'vue'
import { mount, createLocalVue } from '@vue/test-utils'
import { install } from '../install'

export const createWrapper = (template, _data, _options) => {
  const localVue = createLocalVue()
  localVue.use(install)
  return mount(
    {
      data() {
        return {
          ..._data,
        }
      },
      ..._options,
      template: template,
    },
    { localVue, attachTo: document.body }
  )
}

export const sleep = (time) => {
  return new Promise((resolve) => {
    setTimeout(resolve, time)
  })
}

export async function nextTick(times = 5) {
  for (let i = 0; i < times; ++i) {
    await Vue.prototype.$nextTick()
  }
}
