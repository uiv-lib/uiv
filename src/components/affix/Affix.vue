<template>
  <div class="hidden-print">
    <div v-scroll="onScroll" :class="classes" :style="styles">
      <slot></slot>
    </div>
  </div>
</template>

<script>
import scroll from './../../directives/scroll'

export default {
  directives: {
    scroll,
  },
  props: {
    offset: {
      type: Number,
      default: 0,
    },
  },
  data() {
    return {
      affixed: false,
    }
  },
  computed: {
    classes() {
      return {
        affix: this.affixed,
      }
    },
    styles() {
      return {
        top: this.affixed ? this.offset + 'px' : null,
      }
    },
  },
  methods: {
    // from https://github.com/ant-design/ant-design/blob/master/components/affix/index.jsx#L20
    onScroll() {
      // if is hidden don't calculate anything
      if (
        !(
          this.$el.offsetWidth ||
          this.$el.offsetHeight ||
          this.$el.getClientRects().length
        )
      ) {
        return
      }
      // get window scroll and element position to detect if have to be normal or affixed
      const scroll = {}
      const element = {}
      const rect = this.$el.getBoundingClientRect()
      const body = document.body
      const types = ['Top', 'Left']
      types.forEach((type) => {
        const t = type.toLowerCase()
        scroll[t] = window['page' + (type === 'Top' ? 'Y' : 'X') + 'Offset']
        element[t] =
          scroll[t] +
          rect[t] -
          (this.$el['client' + type] || body['client' + type] || 0)
      })
      const fix = scroll.top > element.top - this.offset
      if (this.affixed !== fix) {
        this.affixed = fix
        this.$emit(this.affixed ? 'affix' : 'unfix')
        this.$nextTick(() => {
          this.$emit(this.affixed ? 'affixed' : 'unfixed')
        })
      }
    },
  },
}
</script>
