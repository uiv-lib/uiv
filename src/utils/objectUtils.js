// https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object/assign#Polyfill
export function assign (target, varArgs) {
  if (target === null || target === undefined) {
    throw new TypeError('Cannot convert undefined or null to object')
  }
  let to = Object(target)
  for (let index = 1; index < arguments.length; index++) {
    let nextSource = arguments[index]
    if (nextSource !== null && nextSource !== undefined) {
      for (let nextKey in nextSource) {
        // Avoid bugs when hasOwnProperty is shadowed
        if (Object.prototype.hasOwnProperty.call(nextSource, nextKey)) {
          to[nextKey] = nextSource[nextKey]
        }
      }
    }
  }
  return to
}

export function isExist (obj) {
  return typeof obj !== 'undefined' && obj !== null
}

export function isFunction (obj) {
  return typeof obj === 'function'
}

export function isNumber (obj) {
  return typeof obj === 'number'
}

export function isString (obj) {
  return typeof obj === 'string'
}

export function isBoolean (obj) {
  return typeof obj === 'boolean'
}

export function isPromiseSupported () {
  return typeof window !== 'undefined' && isExist(window.Promise)
}
