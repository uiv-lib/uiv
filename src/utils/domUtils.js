import {isExist} from './objectUtils'

export const EVENTS = {
  MOUSE_ENTER: 'mouseenter',
  MOUSE_LEAVE: 'mouseleave',
  FOCUS: 'focus',
  BLUR: 'blur',
  CLICK: 'click',
  INPUT: 'input',
  KEY_DOWN: 'keydown',
  KEY_UP: 'keyup',
  KEY_PRESS: 'keypress',
  RESIZE: 'resize',
  SCROLL: 'scroll'
}

export const TRIGGERS = {
  CLICK: 'click',
  HOVER: 'hover',
  FOCUS: 'focus',
  HOVER_FOCUS: 'hover-focus',
  OUTSIDE_CLICK: 'outside-click',
  MANUAL: 'manual'
}

export const PLACEMENTS = {
  TOP: 'top',
  RIGHT: 'right',
  BOTTOM: 'bottom',
  LEFT: 'left'
}

export function getViewportSize () {
  let width = Math.max(document.documentElement.clientWidth, window.innerWidth || 0)
  let height = Math.max(document.documentElement.clientHeight, window.innerHeight || 0)
  return {width, height}
}

let scrollbarWidth = null
let savedScreenSize = null

export function getScrollbarWidth (recalculate = false) {
  let screenSize = getViewportSize()
  // return directly when already calculated & not force recalculate & screen size not changed
  if (scrollbarWidth !== null && !recalculate &&
    screenSize.height === savedScreenSize.height && screenSize.width === savedScreenSize.width) {
    return scrollbarWidth
  }
  if (document.readyState === 'loading') {
    return null
  }
  const div1 = document.createElement('div')
  const div2 = document.createElement('div')
  div1.style.width = div2.style.width = div1.style.height = div2.style.height = '100px'
  div1.style.overflow = 'scroll'
  div2.style.overflow = 'hidden'
  document.body.appendChild(div1)
  document.body.appendChild(div2)
  scrollbarWidth = Math.abs(div1.scrollHeight - div2.scrollHeight)
  document.body.removeChild(div1)
  document.body.removeChild(div2)
  // save new screen size
  savedScreenSize = screenSize
  return scrollbarWidth
}

export function on (element, event, handler) {
  element.addEventListener(event, handler)
}

export function off (element, event, handler) {
  element.removeEventListener(event, handler)
}

export function isElement (el) {
  return el && el.nodeType === Node.ELEMENT_NODE
}

export function removeFromDom (el) {
  isElement(el) && isElement(el.parentNode) && el.parentNode.removeChild(el)
}

export function ensureElementMatchesFunction () {
  if (!Element.prototype.matches) {
    Element.prototype.matches =
      Element.prototype.matchesSelector ||
      Element.prototype.mozMatchesSelector ||
      Element.prototype.msMatchesSelector ||
      Element.prototype.oMatchesSelector ||
      Element.prototype.webkitMatchesSelector ||
      function (s) {
        let matches = (this.document || this.ownerDocument).querySelectorAll(s)
        let i = matches.length
        while (--i >= 0 && matches.item(i) !== this) {
        }
        return i > -1
      }
  }
}

export function addClass (el, className) {
  if (!isElement(el)) {
    return
  }
  if (el.className) {
    let classes = el.className.split(' ')
    if (classes.indexOf(className) < 0) {
      classes.push(className)
      el.className = classes.join(' ')
    }
  } else {
    el.className = className
  }
}

export function removeClass (el, className) {
  if (!isElement(el)) {
    return
  }
  if (el.className) {
    let classes = el.className.split(' ')
    let newClasses = []
    for (let i = 0, l = classes.length; i < l; i++) {
      if (classes[i] !== className) {
        newClasses.push(classes[i])
      }
    }
    el.className = newClasses.join(' ')
  }
}

export function hasClass (el, className) {
  if (!isElement(el)) {
    return false
  }
  let classes = el.className.split(' ')
  for (let i = 0, l = classes.length; i < l; i++) {
    if (classes[i] === className) {
      return true
    }
  }
  return false
}

export function setDropdownPosition (dropdown, trigger, options = {}) {
  let doc = document.documentElement
  let containerScrollLeft = (window.pageXOffset || doc.scrollLeft) - (doc.clientLeft || 0)
  let containerScrollTop = (window.pageYOffset || doc.scrollTop) - (doc.clientTop || 0)
  let rect = trigger.getBoundingClientRect()
  let dropdownRect = dropdown.getBoundingClientRect()
  dropdown.style.right = 'auto'
  dropdown.style.bottom = 'auto'
  if (options.menuRight) {
    dropdown.style.left = containerScrollLeft + rect.left + rect.width - dropdownRect.width + 'px'
  } else {
    dropdown.style.left = containerScrollLeft + rect.left + 'px'
  }
  if (options.dropup) {
    dropdown.style.top = containerScrollTop + rect.top - dropdownRect.height - 4 + 'px'
  } else {
    dropdown.style.top = containerScrollTop + rect.top + rect.height + 'px'
  }
}

export function isAvailableAtPosition (trigger, popup, placement) {
  let triggerRect = trigger.getBoundingClientRect()
  let popupRect = popup.getBoundingClientRect()
  let viewPortSize = getViewportSize()
  let top = true
  let right = true
  let bottom = true
  let left = true
  switch (placement) {
    case PLACEMENTS.TOP:
      top = triggerRect.top >= popupRect.height
      left = triggerRect.left + triggerRect.width / 2 >= popupRect.width / 2
      right = triggerRect.right - triggerRect.width / 2 + popupRect.width / 2 <= viewPortSize.width
      break
    case PLACEMENTS.BOTTOM:
      bottom = triggerRect.bottom + popupRect.height <= viewPortSize.height
      left = triggerRect.left + triggerRect.width / 2 >= popupRect.width / 2
      right = triggerRect.right - triggerRect.width / 2 + popupRect.width / 2 <= viewPortSize.width
      break
    case PLACEMENTS.RIGHT:
      right = triggerRect.right + popupRect.width <= viewPortSize.width
      top = triggerRect.top + triggerRect.height / 2 >= popupRect.height / 2
      bottom = triggerRect.bottom - triggerRect.height / 2 + popupRect.height / 2 <= viewPortSize.height
      break
    case PLACEMENTS.LEFT:
      left = triggerRect.left >= popupRect.width
      top = triggerRect.top + triggerRect.height / 2 >= popupRect.height / 2
      bottom = triggerRect.bottom - triggerRect.height / 2 + popupRect.height / 2 <= viewPortSize.height
      break
  }
  return top && right && bottom && left
}

export function setTooltipPosition (tooltip, trigger, placement, auto, appendToSelector) {
  let container
  let containerScrollTop
  let containerScrollLeft
  if (!isExist(appendToSelector) || appendToSelector === 'body') {
    container = document.body
    let doc = document.documentElement
    containerScrollLeft = (window.pageXOffset || doc.scrollLeft) - (doc.clientLeft || 0)
    containerScrollTop = (window.pageYOffset || doc.scrollTop) - (doc.clientTop || 0)
  } else {
    container = document.querySelector(appendToSelector)
    containerScrollLeft = container.scrollLeft
    containerScrollTop = container.scrollTop
  }
  // auto adjust placement
  if (auto) {
    // Try: right -> bottom -> left -> top
    // Cause the default placement is top
    let placements = [PLACEMENTS.RIGHT, PLACEMENTS.BOTTOM, PLACEMENTS.LEFT, PLACEMENTS.TOP]
    // The class switch helper function
    const changePlacementClass = (placement) => {
      // console.log(placement)
      placements.forEach(placement => {
        removeClass(tooltip, placement)
      })
      addClass(tooltip, placement)
    }
    // No need to adjust if the default placement fits
    if (!isAvailableAtPosition(trigger, tooltip, placement)) {
      for (let i = 0, l = placements.length; i < l; i++) {
        // Re-assign placement class
        changePlacementClass(placements[i])
        // Break if new placement fits
        if (isAvailableAtPosition(trigger, tooltip, placements[i])) {
          placement = placements[i]
          break
        }
      }
      changePlacementClass(placement)
    }
  }
  // fix left and top for tooltip
  let rect = trigger.getBoundingClientRect()
  let tooltipRect = tooltip.getBoundingClientRect()
  if (placement === PLACEMENTS.BOTTOM) {
    tooltip.style.top = containerScrollTop + rect.top + rect.height + 'px'
    tooltip.style.left = containerScrollLeft + rect.left + rect.width / 2 - tooltipRect.width / 2 + 'px'
  } else if (placement === PLACEMENTS.LEFT) {
    tooltip.style.top = containerScrollTop + rect.top + rect.height / 2 - tooltipRect.height / 2 + 'px'
    tooltip.style.left = containerScrollLeft + rect.left - tooltipRect.width + 'px'
  } else if (placement === PLACEMENTS.RIGHT) {
    tooltip.style.top = containerScrollTop + rect.top + rect.height / 2 - tooltipRect.height / 2 + 'px'
    tooltip.style.left = containerScrollLeft + rect.left + rect.width + 'px'
  } else {
    tooltip.style.top = containerScrollTop + rect.top - tooltipRect.height + 'px'
    tooltip.style.left = containerScrollLeft + rect.left + rect.width / 2 - tooltipRect.width / 2 + 'px'
  }
}

export function hasScrollbar (el) {
  return el.scrollHeight > el.clientHeight
}

export function toggleBodyOverflow (enable) {
  if (enable) {
    document.body.style.paddingRight = null
  } else {
    if (hasScrollbar(document.documentElement)) {
      document.body.style.paddingRight = `${getScrollbarWidth()}px`
    }
  }
}

export function getClosest (el, selector) {
  ensureElementMatchesFunction()
  let parent
  while (el) {
    parent = el.parentElement
    if (parent && parent.matches(selector)) {
      return parent
    }
    el = parent
  }
  return null
}

export function getParents (el, selector, until = null) {
  ensureElementMatchesFunction()
  let parents = []
  let parent = el.parentElement
  while (parent) {
    if (parent.matches(selector)) {
      parents.push(parent)
    } else if (until && (until === parent || parent.matches(until))) {
      break
    }
    parent = parent.parentElement
  }
  return parents
}
