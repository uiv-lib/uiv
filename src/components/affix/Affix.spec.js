import Affix from './Affix'
import { createWrapper, nextTick } from '../../__test__/utils'

describe('Affix', () => {
  let wrapper

  beforeEach(() => {
    window.scrollTo(0, 0)
    wrapper = createWrapper(
      `
<div>
  <affix :offset="offset">
    <ul>
      <li>Test1</li>
      <li>Test2</li>
      <li>Test3</li>
    </ul>
  </affix>
</div>`,
      {
        offset: 0,
      }
    )
  })

  test('renders correctly', () => {
    expect(wrapper.element).toMatchSnapshot()
  })

  it('should be able to toggle affix class', async () => {
    expect(wrapper.findAll('.affix').length).toEqual(0)
    wrapper.findComponent(Affix).vm.$el.getClientRects = jest.fn(() => 'emt')
    wrapper.findComponent(Affix).vm.$el.getBoundingClientRect = jest.fn(() => ({
      top: -200,
      left: 0,
    }))
    window.scrollTo(0, 200)
    // console.log(document.body.innerHTML)
    await nextTick()
    // console.log(document.body.innerHTML)
    expect(wrapper.findAll('.affix').length).toEqual(1)
    wrapper.findComponent(Affix).vm.$el.getBoundingClientRect = jest.fn(() => ({
      top: -1,
      left: 0,
    }))
    window.scrollTo(0, 500)
    await nextTick()
    expect(wrapper.findAll('.affix').length).toEqual(1)
    wrapper.findComponent(Affix).vm.$el.getBoundingClientRect = jest.fn(() => ({
      top: 0,
      left: 0,
    }))
    window.scrollTo(0, 0)
    await nextTick()
    expect(wrapper.findAll('.affix').length).toEqual(0)
  })

  it('should not toggle affix class if element is hidden', async () => {
    expect(wrapper.findAll('.affix').length).toEqual(0)
    window.scrollTo(0, 500)
    await nextTick()
    expect(wrapper.findAll('.affix').length).toEqual(0)
    window.scrollTo(0, 0)
    await nextTick()
    expect(wrapper.findAll('.affix').length).toEqual(0)
  })

  it('should be able to use offset', async () => {
    wrapper.findComponent(Affix).vm.$el.getClientRects = jest.fn(() => 'emt')
    await wrapper.setData({
      offset: 50,
    })
    wrapper.findComponent(Affix).vm.$el.getBoundingClientRect = jest.fn(() => ({
      top: 60,
      left: 0,
    }))
    window.scrollTo(0, 40)
    await nextTick()
    expect(wrapper.findAll('.affix').length).toEqual(0)
    wrapper.findComponent(Affix).vm.$el.getBoundingClientRect = jest.fn(() => ({
      top: 45,
      left: 0,
    }))
    window.scrollTo(0, 55)
    await nextTick()
    expect(wrapper.findAll('.affix').length).toEqual(1)
  })
})
