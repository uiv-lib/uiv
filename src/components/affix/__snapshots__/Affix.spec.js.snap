// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Affix renders correctly 1`] = `
<div>
  <div
    class="hidden-print"
  >
    <div
      class=""
    >
      <ul>
        <li>
          Test1
        </li>
         
        <li>
          Test2
        </li>
         
        <li>
          Test3
        </li>
      </ul>
    </div>
  </div>
</div>
`;
